src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,244 @@
+                    continue;
+                const auto& expr_input_ports = port_expr->get_input_ports();
+                extract_applicable = true;
+                if (expr_input_ports.size() == 0)
+                    extract_applicable = false;
+                for (size_t i = 0; i < expr_input_ports.size(); ++i) {  // iter expr ports
+                    const auto& parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
+                    bool parent_scalar_with_single_consumer =
+                        ov::is_type<snippets::op::Scalar>(parent->get_node()) &&
+                        parent->get_output_port_connector(0)->get_consumers().size() == 1;
+
+                    const auto& loop_port = inner_loop_info->find_loop_port(expr_input_ports[i]);
+                    if (loop_port == inner_loop_input_ports.end() && !parent_scalar_with_single_consumer) {
+                        // expr input port is not a loop input port, then should not extract
+                        // expr with single scalar parent could be extracted as well, which is common.
+                        extract_applicable = false;
+                        break;
+                    }
+                    if (loop_port != inner_loop_input_ports.end()) {
+                        if (get_stride_after_move_outer(*loop_port) != 1) {
+                            // after move to outside, stride in inner loop is not 1, then should not extract.
+                            extract_applicable = false;
+                            break;
+                        }
+                    }
+                }
+                if (extract_applicable) {
+                    // extract
+                    extracted_exprs.push_back(port_expr);
+                    LinearIR::constExprIt inner_loop_begin_pos, inner_loop_end_pos;
+                    std::tie(inner_loop_begin_pos, inner_loop_end_pos) = loop_manager->get_loop_bounds(linear_ir, inner_loop_id);
+                    for (size_t i = 0; i < port_expr->get_input_count(); ++i) {
+                        auto parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
+                        if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
+                            // update parent scalar loop id
+                            remove_last_loop_id(parent);
+                            // find iterator for parent scalar
+                            auto parent_scalar_iter = std::find(inner_loop_begin_pos, inner_loop_end_pos, parent);
+                            // move if it is not first
+                            if (*parent_scalar_iter != *inner_loop_begin_pos) {
+                                linear_ir.move(parent_scalar_iter, inner_loop_begin_pos);
+                            } else {
+                                inner_loop_begin_pos++;  // keep expr order, scalar is before consumer
+                            }
+                        }
+                    }
```

Comment:
As minor idea:
To create the separate function `extract_scalars(...)` again:
```
void extract_scalars(const ExpressionPtr& expr, LinearIR& linear_ir, LinearIR::constExprIt& inner_loop_begin_pos, LinearIR::constExprIt inner_loop_end_pos) {
    for (size_t i = 0; i < expr->get_input_count(); ++i) {
        auto parent = expr->get_input_port_connector(i)->get_source().get_expr();
        if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
            // update parent scalar loop id
            remove_last_loop_id(parent);
            // find iterator for parent scalar
            auto parent_scalar_iter = std::find(inner_loop_begin_pos, inner_loop_end_pos, parent);
            // move if it is not first
            if (*parent_scalar_iter != *inner_loop_begin_pos) {
                linear_ir.move(parent_scalar_iter, inner_loop_begin_pos);
            } else {
                inner_loop_begin_pos++;  // keep expr order, scalar is before consumer
            }
        }
    }
}
```

Reply:
Another idea is to create the function `extract_node` since you use it twice: for scalar and expressions:

```
void extract_expr(const ExpressionPtr& expr, LinearIR& linear_ir, LinearIR::constExprIt& inner_loop_begin_pos, LinearIR::constExprIt inner_loop_end_pos) {
    // update expr loop id
    remove_last_loop_id(expr);
    // move if it is not the first
    if (expr != *inner_loop_begin_pos) {
        auto port_expr_iter = std::find(inner_loop_begin_pos, inner_loop_end_pos, expr);
        // assert
        linear_ir.move(port_expr_iter, inner_loop_begin_pos);
    } else {
        inner_loop_begin_pos++;
    }
}
....
// extract_from_loop function body:
if (is_extraction_applicable(port_expr, inner_loop_info)) {
    LinearIR::constExprIt inner_loop_begin_pos, inner_loop_end_pos;
    std::tie(inner_loop_begin_pos, inner_loop_end_pos) = loop_manager->get_loop_bounds(linear_ir, inner_loop_id);
    // extract scalar on inputs if there are
    for (size_t i = 0; i < port_expr->get_input_count(); ++i) {
        auto parent = port_expr->get_input_port_connector(i)->get_source().get_expr();
        if (ov::is_type<snippets::op::Scalar>(parent->get_node())) {
            extract_expr(parent, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
        }
    }
    // extract expr
    extract_expr(port_expr, linear_ir, inner_loop_begin_pos, inner_loop_end_pos);
    
    ...
}
```

Reply:
extract_node() is used for both target expr and its scalar. done in [c575707](https://github.com/openvinotoolkit/openvino/pull/24044/commits/c5757078d0c1dbf4e73f824cb21c35a10b6d7cfd). Thanks Alexandra.